fineract-provider/src/main/java/org/apache/fineract/infrastructure/security/api/AuthenticationApiResource.java

```diff
@@ -158,8 +167,54 @@ public String authenticate(@Parameter(hidden = true) final String apiRequestBody
                         returnClientList ? clientReadPlatformService.retrieveUserClients(userId) : null);
             }
 
+            String clientIp = "Unknown IP Address";
+            if (servletRequest != null) {
+                clientIp = servletRequest.getHeader("X-Forwarded-For");
+                if (clientIp == null || clientIp.isEmpty()) {
+                    clientIp = servletRequest.getRemoteAddr();
+                } else {
+                    // The X-Forwarded-For header can contain multiple IP addresses, in case of proxies.
+                    // The first IP in the list is the original client.
+                    clientIp = Iterables.get(Splitter.on(',').split(clientIp), 0);
+                    ;
+                }
+            }
```

Comment:
why is business logic happening at the presentation layer ? I think this should be moved to the service layer